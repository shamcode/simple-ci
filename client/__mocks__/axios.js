const METHODS = [
    'get',
    'post',
    'put',
    'delete'
];

const FIXTURES = {};
const MOCKS = {};

function mockMethodFactory( method ) {
    return jest.fn().mockImplementation(
        ( url ) => {
            if ( !FIXTURES[ method ].hasOwnProperty( url ) ) {
                throw new Error( `Missing fixture for ${method.toUpperCase()} url = "${url}"` );
            }
            const { success, data } = FIXTURES[ method ][ url ];
            return Promise[ success ? 'resolve' : 'reject' ]( {
                data
            } );
        }
    );
}

METHODS.forEach(
    method => {
        MOCKS[ method ] = mockMethodFactory( method );
        FIXTURES[ method ] = {};
    }
);

const defaultMocksData = ( () => {
    const data = {};
    data.chain = {
        id: 2,
        name: 'Test chain',
        command: 'echo 1'
    };
    data.project = {
        id: 1,
        name: 'Test',
        cwd: '/tmp/',
        chains: [ data.chain ]
    };
    return data;
} )();

export default {
    defaultMocksData: defaultMocksData,
    mocks: MOCKS,
    create: jest.fn().mockImplementation( () => {
        return {
            ...MOCKS,
            interceptors: {
                request: {
                    use: () => {}
                },
                response: {
                    use: () => {}
                }
            }
        };
    } ),
    use( method, url, data, success = true ) {
        if ( !METHODS.includes( method ) ) {
            throw new Error( `Unknown method: "${method}"`);
        }
        FIXTURES[ method ][ url ] = {
            data,
            success
        };
        return this;
    },
    useDefaultMocks() {
        this
            .use( 'get', '/projects', [ defaultMocksData.project ] )
            .use( 'get', '/projects/1', defaultMocksData.project )
            .use( 'get',  '/chains/2', defaultMocksData.chain );
        return this;
    }
};