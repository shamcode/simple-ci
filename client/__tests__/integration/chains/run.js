import setup, { app } from '../helpers'
import axios from 'axios';
jest.mock( 'axios' );

beforeEach( () => {
    jest.resetModules();
    jest.clearAllMocks();
    setup();
} );


it( 'run project chain', async () => {
    expect.assertions( 6 );

    const wsMock = {
        send: jest.fn()
    };
    window.WebSocket = jest.fn( () => wsMock );
    axios
        .useDefaultMocks()
        .use( 'post', '/chains/2/run', null );

    await app.start();
    await app.project.open();
    await app.chain.open();
    await app.click( '[href="projects/1/chains/2/run"]' );

    await app.waitRendering();
    app.checkBody();

    await app.click( '.run-project-chain' );
    wsMock.onopen();
    await app.waitRendering();

    expect( wsMock.send.mock.calls.length ).toBe( 1 );
    expect( wsMock.send.mock.calls[ 0 ][ 0 ] ).toBe( '{"chainId":2}' );
    expect( axios.mocks.post ).toHaveBeenCalledTimes( 1 );
    expect( axios.mocks.post.mock.calls[ 0 ][ 0 ] ).toBe( '/chains/2/run' );
    app.checkBody();
} );

