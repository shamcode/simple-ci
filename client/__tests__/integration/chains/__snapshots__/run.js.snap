// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`run project chain 1`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI
    <!--default-->
  </a>
</header>
<!--Header-->
<main>
  <ul class=\\"breadcrumb\\">
    <li>
      <a href=\\"\\" class=\\"\\">Projects
        <!--default-->
      </a>
    </li>
    <li>
      <a href=\\"projects/1\\" class=\\"\\">Test
        <!--default-->
      </a>
    </li>
    <!--if-->
    <li>
      <a href=\\"projects/1/chains/2\\" class=\\"\\">Test chain
        <!--default-->
      </a>
    </li>
    <!--if-->
    <li>Run chain</li>
  </ul>
  <!--BreadCrumb-->
  <!--breadcrumbs-->
  <!--if-->
  <!--ErrorsPanel-->
  <button class=\\"run-project-chain\\"><span>Run</span></button> <pre class=\\"command-output\\"></pre>
  <!--CommandOutput-->
  <!--content-->
  <!--if-->
  <!--PageContent-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;

exports[`run project chain 2`] = `
"<header>
  <a href=\\"\\" class=\\"logo\\">Simple CI
    <!--default-->
  </a>
</header>
<!--Header-->
<main>
  <ul class=\\"breadcrumb\\">
    <li>
      <a href=\\"\\" class=\\"\\">Projects
        <!--default-->
      </a>
    </li>
    <li>
      <a href=\\"projects/1\\" class=\\"\\">Test
        <!--default-->
      </a>
    </li>
    <!--if-->
    <li>
      <a href=\\"projects/1/chains/2\\" class=\\"\\">Test chain
        <!--default-->
      </a>
    </li>
    <!--if-->
    <li>Run chain</li>
  </ul>
  <!--BreadCrumb-->
  <!--breadcrumbs-->
  <!--if-->
  <!--ErrorsPanel-->
  <button class=\\"run-project-chain\\"><span>Run</span></button> <pre class=\\"command-output\\"></pre>
  <!--CommandOutput-->
  <!--content-->
  <!--if-->
  <!--PageContent-->
</main>
<footer> Sham 2018 </footer>
<!--Footer-->"
`;
